include/arch/arc/arch.h

```diff
@@ -89,55 +89,66 @@ extern "C" {
 		1 << (31 - __builtin_clz(x) + 1) : \
 		1 << (31 - __builtin_clz(x)))
 
-
-#if defined(CONFIG_USERSPACE)
-#define ARCH_THREAD_STACK_RESERVED \
-	(STACK_GUARD_SIZE + CONFIG_PRIVILEGED_STACK_SIZE)
-#else
-#define ARCH_THREAD_STACK_RESERVED (STACK_GUARD_SIZE)
-#endif
-
-
-#if  defined(CONFIG_USERSPACE) && CONFIG_ARC_MPU_VER == 2
+#if CONFIG_ARC_MPU_VER == 2
 /* MPUv2 requires
  *  - region size must be power of 2 and >= 2048
  *  - region start must be aligned to its size
  */
-#define Z_ARC_MPUV2_SIZE_ALIGN(size) POW2_CEIL(STACK_SIZE_ALIGN(size))
+#define Z_ARC_MPU_SIZE_ALIGN(size) POW2_CEIL(STACK_SIZE_ALIGN(size))
+#elif CONFIG_ARC_MPU_VER == 3
+/* MPUv3 requires
+ * - region size must be 32 bytes aligned
+ * - region start must be 32 bytes aligned
+ */
+#define Z_ARC_MPU_SIZE_ALIGN(size) STACK_SIZE_ALIGN(size)
+#endif
+
+
+
+#if  defined(CONFIG_USERSPACE)
+
+#if CONFIG_ARC_MPU_VER == 2
+/* MPU stack guard does not works for MPUv2 as it uses GEN_PRIV_STACK */
+#define ARCH_THREAD_STACK_RESERVED 0
+#define Z_PRIVILEGE_STACK_ALIGN CONFIG_PRIVILEGED_STACK_SIZE
 /*
- * user stack and guard are protected using MPU regions, so need to adhere to
+ * user stack are protected using MPU regions, so need to adhere to
  * MPU start, size alignment
  */
-#define Z_ARC_THREAD_STACK_ALIGN(size)	Z_ARC_MPUV2_SIZE_ALIGN(size)
+#define Z_ARC_THREAD_STACK_ALIGN(size)	Z_ARC_MPU_SIZE_ALIGN(size)
+#define ARCH_THREAD_STACK_LEN(size) Z_ARC_MPU_SIZE_ALIGN(size)
+
+#elif CONFIG_ARC_MPU_VER == 3
+#define ARCH_THREAD_STACK_RESERVED \
+		(STACK_GUARD_SIZE + CONFIG_PRIVILEGED_STACK_SIZE)
+#define Z_PRIVILEGE_STACK_ALIGN (STACK_ALIGN)
+
+#define Z_ARC_THREAD_STACK_ALIGN(size) (STACK_ALIGN)
 #define ARCH_THREAD_STACK_LEN(size) \
-	(Z_ARC_MPUV2_SIZE_ALIGN(size) + ARCH_THREAD_STACK_RESERVED)
-/*
- * for stack array, each array member should be aligned both in size
- * and start
- */
-#define Z_ARC_THREAD_STACK_ARRAY_LEN(size) \
-		(Z_ARC_MPUV2_SIZE_ALIGN(size) + \
-		MAX(Z_ARC_MPUV2_SIZE_ALIGN(size), \
-		POW2_CEIL(ARCH_THREAD_STACK_RESERVED)))
-#else
+		(Z_ARC_MPU_SIZE_ALIGN(size) + ARCH_THREAD_STACK_RESERVED)
+
+#endif /* CONFIG_ARC_MPU_VER == 2 */
+
+#else /* CONFIG_USERSPACE */
 /*
- * MPUv3, no-mpu and no USERSPACE share the same macro definitions.
  * For MPU STACK_GUARD  kernel stacks do not need a MPU region to protect,
  * only guard needs to be protected and aligned. For MPUv3, MPU_STACK_GUARD
  * requires start 32 bytes aligned, also for size which is decided by stack
  * array and USERSPACE; For MPUv2, MPU_STACK_GUARD requires
```

Comment:
Is MPU stack guard for no-userspace supported with MPU v2?

Reply:
stack guards have always been orthogonal to userspace, so I expect the answer is no.

Reply:
no sense to have MPU stack guard for MPUv2, even no-userspace. Indeed, stack guards have always been orthogonal to userspace, but because of the hardware limitation of ARC MPUv2 (memory waste),  better to drop it, we can use ARC STACK CHECKING instead.

Reply:
@vonhust agreed, 2K of dead memory per thread is not a practical configuration.